packages/merkle-trees/src/sparse_merkle.nr

```diff
@@ -1,112 +1,208 @@
-use crate::{MembershipProver, Modifier, NonMembershipProver, SMT_Creator};
+pub use crate::types::{Calculator, MembershipProver, Modifier, NonMembershipProver, SMT_Creator};
 
-mod tests;
-mod tree;
-
-pub struct SparseMerkleTree {
-    root: Field,
-    leaf_hasher: fn([Field; 3]) -> Field,
-    hasher: fn([Field; 2]) -> Field,
+pub struct SparseMerkleTree<T> {
+    pub root: T,
+    pub leaf_hasher: fn([T; 3]) -> T,
+    pub hasher: fn([T; 2]) -> T,
 }
 
-impl SMT_Creator for SparseMerkleTree {
-    fn default(
-        root: Field,
-        leaf_hasher: fn([Field; 3]) -> Field,
-        hasher: fn([Field; 2]) -> Field,
-    ) -> Self {
+impl<T> SMT_Creator<T> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
+    fn default(root: T, leaf_hasher: fn([T; 3]) -> T, hasher: fn([T; 2]) -> T) -> Self {
         Self { root, leaf_hasher, hasher }
     }
-}
 
-impl MembershipProver<(Field, Field), Field, Field> for SparseMerkleTree {
-    fn membership<let N: u32>(self, entry: (Field, Field), indexes: Field, hash_path: [Field; N]) {
-        let mut calculated_root: Field = 0;
+    fn from(root: T, leaf_hasher: fn([T; 3]) -> T, hasher: fn([T; 2]) -> T) -> Self {
+        Self::default(root, leaf_hasher, hasher)
+    }
+
+    fn new(leaf_hasher: fn([T; 3]) -> T, hasher: fn([T; 2]) -> T) -> Self {
+        Self::from(T::default(), leaf_hasher, hasher)
+    }
+}
 
+impl<T> MembershipProver<T, (T, T)> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
+    fn membership<let N: u32>(self, entry: (T, T), indexes: Field, hash_path: [T; N]) {
         // membership proof: the root is calculated based on the entry, the siblings,
         // and the path determined by the key of entry through consecutive hashing
-        calculated_root = self.calculate_root(entry, indexes, hash_path);
-
-        assert(calculated_root == self.root);
+        assert(self.calculate_root(entry, indexes, hash_path) == self.root);
     }
 }
 
-impl NonMembershipProver for SparseMerkleTree {
+impl<T> NonMembershipProver<T, (T, T)> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
     fn non_membership(
         self,
-        entry: (Field, Field),
-        matching_entry: (Field, Field),
-        siblings: [Field; 256],
+        entry: (T, T),
```

Comment:
Since `index` is used rather than the key of entry, might as well remove the `entry` argument as it can make it confusing to have both

Reply:
Good point, will remove, thanks!